File: src/main/java/com/leetcode/FourFour.java
package com.leetcode;

/**
 * @author: John
 * @date: 2020-03-15 16:18
 * @description: 44
 **/
public class FourFour {

//    public boolean comparison(String str, String pattern) {
//        int s = 0, p = 0, match = 0, starIdx = -1;
//        while (s < str.length()) {
//            // advancing both pointers
//            if (p < pattern.length() && (pattern.charAt(p) == '?' ||
//                    str.charAt(s) == pattern.charAt(p))) {
//                s++;
//                p++;
//            }
//            // found only advancing pattern pointer
//            else if (p < pattern.length() && pattern.charAt(p) == '*') {
//                starIdx = p;
//                match = s;
//                p++;
//            }
//            // last pattern pointer was *, advancing string pointer
//            else if (starIdx != -1) {
//                p = starIdx + 1;
//                match++;
//                s = match;
//            }
//            // current pattern pointer is not star, last patter pointer was not *
//            // characters do not match
//            else {
//                return false;
//            }
//        }
//        // check for remaining characters in pattern
//        while (p < pattern.length() && pattern.charAt(p) == '*') {
//            p++;
//        }
//        return p == pattern.length();
//    }
//    public boolean isMatch(String s, String p) {
//       boolean[][] match = new boolean[s.length() + 1][p.length() + 1];
//       match[s.length()][p.length()] = true;
//       for (int i = p.length() - 1; i >= 0; i--) {
//           if (p.charAt(i)!='*') {
//               break;
//           } else {
//               match[s.length()][i] = true;
//           }
//       }
//       for (int i = s.length() - 1; i >= 0; i--) {
//           for (int j = p.length() - 1; j >= 0; j--) {
//               if (s.charAt(i) == p.charAt(j) || p.charAt(j) == '?') {
//                   match[i][j] = match[i+1][j+1];
//               } else if (p.charAt(j) == '*') {
//                   match[i][j] = match[i][j+1] || match[i + 1][j];
//               } else {
//                   match[i][j] = false;
//               }
//           }
//       }
//       return match[0][0];
//    }

//    public boolean isMatch_2d_method(String s, String p) {
//       int m = s.length(), n = p.length();
//       boolean[][] dp = new boolean[m+1][n+1];
//       dp[0][0] = true;
//       for (int i = 1; i <= m; i++) {
//           dp[i][0] = false;
//       }
//
//       for (int j = 1; j <= n; j++) {
//           if (p.charAt(j - 1) == '*') {
//               dp[0][j] = true;
//           } else {
//               break;
//           }
//       }
//
//       for (int i = 1; i <= m; i++) {
//           for (int j = 1; j <= n; j++) {
//               if (p.charAt(j-1) != '*') {
//                   dp[i][j] = dp[i-1][j-1] && (s.charAt(i-1) == p.charAt(j-1) ||
//                           p.charAt(j - 1) == '?');
//               } else {
//                   dp[i][j] = dp[i-1][j] || dp[i][j-1];
//               }
//           }
//       }
//       return dp[m][n];
//    }

//    public boolean isMatch(String s, String p) {
//       int m = s.length(), n = p.length();
//       boolean[] dp = new boolean[n];
//       dp[0] = true;
//       for (int i = 1; i <= m; i++) {
//           for (int j = 1; j <= n; j++) {
//               if () {
//
//               }
//           }
//       }
//    }


    public static void main(String[] args) {
        FourFour fourfour = new FourFour();
//        fourfour.comparison("adceb", "*a*b");
//        fourfour.comparison("aa", "*");
    }
}

File: leetcodecpp/leetcode/FourFour.cpp
//
// Created by John on 2020/3/15.
//

#include <iostream>
#include <vector>
#include "string"

using namespace std;
class Solution{
//public:
//    bool isMatch(string s, string p) {
//        if (p.empty() || s.empty()) {
//            return s.empty() && p.empty();
//        }
//        bool first_match = (!s.empty() && (s[0] == p[0] || p[0] == '?' ||
//                p[0] == '*' && s[0] == p[1]));
//        if (p[0] == '*') {
//            if (p[1] == '*') {
//                return isMatch(s, p.substr(1));
//            }
//            return isMatch(s.substr(1), p);
//        } else {
//            return first_match &&  (isMatch(s.substr(1), p.substr(1)));
//        }
//    }
//public:
//    bool isMatch(const char *s, const char *p) {
//        const char* star = NULL;
//        const char* ss = s;
//        while (*s) {
//            // advancing both pointers when(both characters match) or ('?' found in pattern)
//            // note that *p will not advance beyond its length
//            if ((*p == '?') || (*p==*s)) {
//                s++;p++;
//                continue;
//            }
//            // * found in pattern, track index of *, only advancing pattern pointer
//            if (*p == '*') {
//                star = p++;
//                ss=s;
//                continue;
//            }
//            // current characters didn't match, last pattern pointer was *, current pattern pointer is not *
//            // only advancing pattern pointer
//            if (star) {
//                p = star + 1;
//                s = ++ss;
//                continue;
//            }
//            // current pattern pointer is not star, last patter pointer was not *
//            // characters do not match
//            return false;
//        }
//        // check for remaining characters in pattern
//        while (*p == '*') {
//            p++;
//        }
//        return !*p;
//    }


//public:
//    bool isMatch(string s, string p) {
//        int ss = 0, pi = 0, match = 0, startIdx = -1;
//        while (ss < s.size()) {
//            if (pi < p.size() && (p[pi] == '?' || s[ss] == p[pi])) {
//                ss++;
//                pi++;
//            } else if (pi < p.size() && p[pi] == '*') {
//                match = ss;
//                startIdx = pi++;
//            } else if (startIdx != -1) {
//                ss = ++match;
//                pi = startIdx + 1;
//            } else {
//                return false;
//            }
//        }
//        while (pi < p.size() && p[pi] == '*') {
//            pi++;
//        }
//        return pi == p.size();
//    }

public:
    bool isMatch(string s, string p) {
        int m = s.length(), n = p.length();
        if (n >30000) {
            return false;
        }
        vector<bool> cur(m + 1, false);
        cur[0] = true;
        for (int j = 1; j <= n; j++) {
            bool pre = cur[0];
            cur[0] = cur[0] && p[j-1] == '*';
            for (int i = 1; i <= m ; i++) {
                bool temp = cur[i];
                if (p[j-1] != '*') {
                    cur[i] = pre && (s[i-1] == p[j-1] || p[j-1] == '?');
                } else {
                    cur[i] = cur[i-1] || cur[i];
                }
                pre = temp;
            }
        }
        return cur[m];
    }

};

//int main() {
//    Solution * solution = new Solution();
////    cout << solution->isMatch("acdcb", "a*c?b") << endl;
////    cout << solution->isMatch("acdcb", "a*c?b") << endl;
//    cout << solution->isMatch("adceb", "*a*b") << endl;
//}